azkaban-common/src/main/java/azkaban/project/FlowTrigger.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2017 LinkedIn Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.project;
+
+import azkaban.Constants;
+import azkaban.utils.Props;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import org.apache.commons.lang.StringUtils;
+
+/**
+ * FlowTrigger is an immutable class which holds
+ * all the data and properties of a flow trigger.
+ */
+
+public class FlowTrigger {
+
+  private final List<FlowTriggerDependency> dependencies;
+  private final Props props; // trigger level props
+
+  private FlowTrigger(final Props props, final List<FlowTriggerDependency> dependencies) {
+    validateProps(props);
+    validateDependencies(dependencies);
+    this.props = new Props(props.getParent(), props);
+    this.dependencies = Collections.unmodifiableList(dependencies);
+  }
+
+  private void validateProps(final Props props) {
```

Comment:
Does it make more sense to have the builder class do the validation?

Reply:
you mean validate inside build()? What's the benefit?

Reply:
Yes.
Since the constructor is private, the only place where an instance can be created is in the builder. It makes more logical sense to me to have the builder validates inputs. This way we can keep the value class itself simpler.